File: DP(Dynamic_Programming, 동적계획법)/백준_9095_1,2,3더하기/1,2,3 더하기.java
import java.util.*;

public class Main {
    static int[] arr = new int[11];

    public static void main(String[] args) {
        arr[1] = 1; arr[2] = 2; arr[3] = 4;
        Scanner s = new Scanner(System.in);
        int T = s.nextInt();
        for(int i = 0; i<T; i++){
            int n = s.nextInt();
            if(arr[n] > 0) System.out.println(arr[n]);
            else{
                for(int j = 4; j <=n; j++){
                    arr[j] = arr[j-1] + arr[j-2] + arr[j-3];
                }
                System.out.println(arr[n]);
            }
        }

    }
}

File: DP(Dynamic_Programming, 동적계획법)/백준_9095_1,2,3더하기/1,2,3 더하기.kt
import java.util.*
import kotlin.math.*

fun main(args: Array<String>){
    val arr = IntArray(11)
    arr[1] = 1; arr[2] = 2; arr[3] = 4;
    val s = Scanner(System.`in`)
    val T = s.nextInt()
    for(i in 1..T){
        val n = s.nextInt()
        if(arr[n] > 0) println("${arr[n]}")
        else{
            for(j in 4..n){
                arr[j] = arr[j-1] + arr[j-2] + arr[j-3];
            }
            println(arr[n])
        }
    }
}